dimod/response.py

```diff
@@ -347,31 +350,170 @@ def from_pandas(cls, samples_df, data_vectors, vartype=None, info=None):
+                response = dimod.Response.from_futures((future,), dimod.BINARY, 3,
+                                                       data_vector_keys={'en': 'energy'})
+
+                future.set_result({'samples': [0, 1, 0], 'en': [1]})
+
+                # now read from the response
+                matrix = response.samples_matrix
+
+        Notes:
+            The future objects are read on the first read of :attr:`.Response.samples_matrix` or
+            :attr:`.Response.data_vectors`.
+
         """
-        if not futures:
-            raise ValueError("Empty list of futures given")
+
+        if data_vector_keys is None:
+            data_vector_keys = {'energy': 'energy'}
+        elif isinstance(data_vector_keys, Mapping):
+            data_vector_keys = dict(data_vector_keys)
+        else:
+            data_vector_keys = {key: key for key in data_vector_keys}  # identity mapping
+
+        if info_keys is None:
+            info_keys = {}
+        elif isinstance(info_keys, Mapping):
+            info_keys = dict(info_keys)
+        else:
+            info_keys = {key: key for key in info_keys}
+
+        # now initialize self with a 0 samples, but the correct number of variables/labels/etc
+        empty_samples_matrix = np.matrix(np.empty((0, num_variables), dtype=np.int8))
+
+        empty_data_vectors = {key: [] for key in itervalues(data_vector_keys)}
+
+        if active_variables is not None:
+            if variable_labels is None:
+                variable_labels = active_variables
+            elif len(variable_labels) != len(active_variables):
+                raise ValueError("active_variables and variable_labels should have the same length")
+
+
+        # let Response parse vartype, variable_labels. We also want to start with an empty info
+        response = Response(samples_matrix=empty_samples_matrix, data_vectors=empty_data_vectors,
+                            vartype=vartype, variable_labels=variable_labels)
+
+        # now dump all of the remaining information into the _futures
+        response._futures = {'futures': futures,
+                             'samples_key': samples_key,
+                             'data_vector_keys': data_vector_keys,
+                             'info_keys': info_keys,
+                             'variable_labels': variable_labels,
+                             'active_variables': active_variables}
+
+        return response
+
+    def _resolve_futures(self, futures, samples_key, data_vector_keys, info_keys,
+                         variable_labels, active_variables):
+
+        # first reset the futures to avoid recursion errors
+        self._futures = {}
 
         # `dwave.cloud.qpu.computation.Future` is not yet interchangeable with
         # `concurrent.futures.Future`, so we need to detect the kind of future
         # we're dealing with.
+        futures = list(futures)  # if generator
         if hasattr(futures[0], 'as_completed'):
```

Comment:
We should check if `futures` is an empty list before accessing `[0]`. In the previous version there was a check, but I don't see it here.